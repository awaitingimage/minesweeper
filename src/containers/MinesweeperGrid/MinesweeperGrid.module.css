.grid {
  display: flex;
  justify-content: center;
  margin: 30px;
  border-spacing: 0px;
}

.grid td {
  border: 2px solid blue;
}
