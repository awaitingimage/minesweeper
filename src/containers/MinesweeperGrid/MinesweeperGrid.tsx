import * as React from 'react';
import styles from './MinesweeperGrid.module.css';
import Tile, { iTile } from '../../components/Tile';

export interface MinesweeperGridProps {
  numOfRows: number;
  numOfColumns: number;
  numOfMines: number;
}

export interface MinesweeperGridState {
  grid: iTile[][];
  currentMines: number[][];
  numOfLoops: number;
  gameFinished: boolean;
}

class MinesweeperGrid extends React.Component<MinesweeperGridProps, MinesweeperGridState> {
  state = { grid: [] as iTile[][], currentMines: [] as number[][], numOfLoops: 0, gameFinished: false };

  generateGrid = () => {
    let grid = [];
    let numOfLoops = this.state.numOfLoops;
    for (let rowIndex = 0; rowIndex < this.props.numOfRows; rowIndex++) {
      let row = [];
      for (let colIndex = 0; colIndex < this.props.numOfColumns; colIndex++) {
        row.push({ x: colIndex, y: rowIndex, value: 0, hidden: true, mine: false, flagged: false });
        numOfLoops++;
      }
      grid.push(row);
    }
    this.setState({ grid, numOfLoops }, this.generateMines);
  };

  generateMines = () => {
    let currentMines: number[][] = [];

    while (currentMines.length < this.props.numOfMines) {
      let newCoordinate = this.generateRandomCoordinate(this.props.numOfRows, this.props.numOfColumns);
      let includesMine = currentMines.filter(coordinate => newCoordinate.toString() === coordinate.toString());

      if (includesMine.length < 1) {
        currentMines.push(newCoordinate);
      }
    }
    this.setState({ currentMines }, this.populateMines);
  };

  populateMines = () => {
    let grid = this.cloneGrid();
    this.state.currentMines.forEach(coordinates => {
      grid[coordinates[0]][coordinates[1]].mine = true;
    });
    this.setState({ grid }, this.generateClues);
  };

  generateClues = () => {
    let grid = this.cloneGrid();
    this.state.currentMines.forEach(coordinates => {
      grid = this.checkNeighbours(grid, coordinates[0], coordinates[1], this.incrementNeighbour);
    });
    this.setState({ grid });
  };

  cloneGrid = () => {
    // look into immutability helpers https://reactjs.org/docs/update.html
    return this.state.grid.map(arr =>
      arr.map(obj => {
        return { ...obj };
      }),
    );
  };

  checkNeighbours = (grid: iTile[][], rowIndex: number, colIndex: number, callback: any) => {
    for (let y = -1; y <= 1; y++) {
      for (let x = -1; x <= 1; x++) {
        this.setState(state => {
          return { numOfLoops: state.numOfLoops + 1 };
        });
        let yIndex = rowIndex + y;
        let xIndex = colIndex + x;
        let samePosition = xIndex === rowIndex && yIndex === colIndex;
        if (
          xIndex >= 0 &&
          xIndex < this.props.numOfColumns &&
          yIndex >= 0 &&
          yIndex < this.props.numOfRows &&
          !samePosition
        ) {
          grid = callback(grid, yIndex, xIndex);
        }
      }
    }
    return grid;
  };

  incrementNeighbour = (grid: iTile[][], yIndex: number, xIndex: number) => {
    if (!grid[yIndex][xIndex].mine) {
      grid[yIndex][xIndex].value++;
    }
    return grid;
  };

  revealNeighbour = (grid: iTile[][], yIndex: number, xIndex: number) => {
    let neighbourTile = grid[yIndex][xIndex];
    if (neighbourTile.value >= 0 && neighbourTile.hidden && !neighbourTile.mine) {
      neighbourTile.hidden = false;
      grid[yIndex][xIndex] = neighbourTile;
      if (neighbourTile.value === 0) grid = this.checkNeighbours(grid, yIndex, xIndex, this.revealNeighbour);
    }
    return grid;
  };

  generateRandomCoordinate = (colMax: number, rowMax: number): number[] => {
    let y = Math.floor(Math.random() * rowMax);
    let x = Math.floor(Math.random() * colMax);
    return [x, y];
  };

  componentDidMount() {
    this.generateGrid();
  }

  componentDidUpdate(prevProps: MinesweeperGridProps) {
    if (JSON.stringify(this.props) !== JSON.stringify(prevProps)) {
      console.log('regenerate grid');
      this.generateGrid();
    }
  }

  tileClicked = (tile: iTile, explosion = false, checkNeighbours = false) => {
    if (explosion) console.log('dead');
    let grid = this.cloneGrid();
    grid[tile.y][tile.x] = tile;
    if (checkNeighbours) {
      grid = this.checkNeighbours(grid, tile.y, tile.x, this.revealNeighbour);
    }
    this.setState({ grid });
  };

  render() {
    return (
      <table className={styles.grid}>
        <tbody>
          {this.state.grid.map((row, rowIndex) => {
            return (
              <tr key={`${this.props.numOfColumns}-${this.props.numOfRows}-${rowIndex}`}>
                {row.map((tile, colIndex) => (
                  <Tile
                    key={`${this.props.numOfColumns}-${this.props.numOfRows}-${rowIndex}-${colIndex}`}
                    tileData={tile}
                    onClick={this.tileClicked}
                  />
                ))}
              </tr>
            );
          })}
        </tbody>
      </table>
    );
  }
}

export default MinesweeperGrid;
