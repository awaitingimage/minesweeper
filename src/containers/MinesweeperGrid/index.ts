import MinesweeperGrid from "./MinesweeperGrid";

export default MinesweeperGrid;
