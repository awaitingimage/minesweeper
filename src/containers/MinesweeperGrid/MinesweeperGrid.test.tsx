import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import MinesweeperGrid from './MinesweeperGrid';

let wrapper: ShallowWrapper<{}, {}, MinesweeperGrid>;
let mineCoordinates = [
  [0, 1],
  [1, 0],
  [1, 1],
  [2, 2],
];
let mineNumber = 0;

const generateRandomCoordinateMock = (colMax: number, rowMax: number): number[] => {
  return mineCoordinates[mineNumber++];
};

let mockFn = jest.fn(generateRandomCoordinateMock);

beforeAll(() => {
  wrapper = shallow(<MinesweeperGrid numOfRows={4} numOfColumns={4} numOfMines={4} />, {
    disableLifecycleMethods: true,
  });
});

it('randomCoordinate returns an array', () => {
  const coordinate = wrapper.instance().generateRandomCoordinate(0, 0);
  expect(Array.isArray(coordinate)).toBe(true);
  expect(coordinate).toStrictEqual([0, 0]);
});

it('called generateCoordinate once', () => {
  wrapper.instance().generateRandomCoordinate = mockFn;
  wrapper.instance().generateGrid();
  expect(mockFn).toHaveBeenCalledTimes(4);
});

it('to match snapshot', () => {
  //const wrapper = shallow(<MinesweeperGrid numOfRows={4} numOfColumns={4} numOfMines={4} />);
  expect(wrapper).toMatchSnapshot();
});
