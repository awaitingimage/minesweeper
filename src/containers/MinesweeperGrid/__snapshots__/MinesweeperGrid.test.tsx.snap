// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`to match snapshot 1`] = `
<table
  className="grid"
>
  <tbody>
    <tr
      key="4-4-0"
    >
      <Tile
        key="4-4-0-0"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 3,
            "x": 0,
            "y": 0,
          }
        }
      />
      <Tile
        key="4-4-0-1"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": true,
            "value": 0,
            "x": 1,
            "y": 0,
          }
        }
      />
      <Tile
        key="4-4-0-2"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 2,
            "x": 2,
            "y": 0,
          }
        }
      />
      <Tile
        key="4-4-0-3"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 0,
            "x": 3,
            "y": 0,
          }
        }
      />
    </tr>
    <tr
      key="4-4-1"
    >
      <Tile
        key="4-4-1-0"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": true,
            "value": 0,
            "x": 0,
            "y": 1,
          }
        }
      />
      <Tile
        key="4-4-1-1"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": true,
            "value": 0,
            "x": 1,
            "y": 1,
          }
        }
      />
      <Tile
        key="4-4-1-2"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 3,
            "x": 2,
            "y": 1,
          }
        }
      />
      <Tile
        key="4-4-1-3"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 1,
            "x": 3,
            "y": 1,
          }
        }
      />
    </tr>
    <tr
      key="4-4-2"
    >
      <Tile
        key="4-4-2-0"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 2,
            "x": 0,
            "y": 2,
          }
        }
      />
      <Tile
        key="4-4-2-1"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 3,
            "x": 1,
            "y": 2,
          }
        }
      />
      <Tile
        key="4-4-2-2"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": true,
            "value": 0,
            "x": 2,
            "y": 2,
          }
        }
      />
      <Tile
        key="4-4-2-3"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 1,
            "x": 3,
            "y": 2,
          }
        }
      />
    </tr>
    <tr
      key="4-4-3"
    >
      <Tile
        key="4-4-3-0"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 0,
            "x": 0,
            "y": 3,
          }
        }
      />
      <Tile
        key="4-4-3-1"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 1,
            "x": 1,
            "y": 3,
          }
        }
      />
      <Tile
        key="4-4-3-2"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 1,
            "x": 2,
            "y": 3,
          }
        }
      />
      <Tile
        key="4-4-3-3"
        onClick={[Function]}
        tileData={
          Object {
            "flagged": false,
            "hidden": true,
            "mine": false,
            "value": 1,
            "x": 3,
            "y": 3,
          }
        }
      />
    </tr>
  </tbody>
</table>
`;
